### DISCUSSION TITLE
[HMR] webpack-dev-server. devServer.watchFiles property breaks hot module replacement

### DISCUSSION THREAD

**Discussion by codewizard13** [webpack/webpack#15250] [Q&A]
Hot module reload was working until I tried to add the `devServer.before` property in webpack.config.js. I researched it, even finding this discussion here (https://github.com/webpack/webpack/discussions/15247) but have found no solution yet. The page HTML page updates fine, but it does a FULL PAGE RELOAD when editing CSS instead of hot replacement. Thanks in advance for your help!

Here is my **webpack.config.js**:

```
const path = require('path')

const postCSSPlugins = [
    require('postcss-import'),
    require('postcss-simple-vars'),
    require('postcss-nested'),
    require('autoprefixer')
]

module.exports = {
    entry: './app/assets/scripts/App.js',
    output: {
        filename: 'bundled.js',
        path: path.resolve(__dirname, 'app')
    },
    devServer: {
        watchFiles: ('./app/**/*.html'),
        static: path.join(__dirname, 'app'),
        hot: true,
        port: 3000,
        // liveReload: false
    },
    mode: 'development',
    module: {
        rules: [
            {
                test: /\.css$/i,
                use: ['style-loader','css-loader', {loader: "postcss-loader", options: {postcssOptions: {plugins: postCSSPlugins}}}]
            }
        ]
    }
}
```

**devServer** property:

```
devServer: {
  before: function(app, server) {
   server._watch('./app/**/*.html')
  },
  static: path.join(__dirname, 'app'),
  hot: true,
  port: 3000
},

```
and my **App.js** file:

```
import '../styles/styles.css';

if (module.hot) {
    module.hot.accept();
}
```
When I tried to add the `before` property I got the error that this property was invalid also. I found this migration guide (https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md) saying that before was replaced with `onBeforeSetupMiddleware` and arguments were changed also. I tried replacing before with this:

**onBeforeSetupMiddleware** property:

```
onBeforeSetupMiddleware: function (devServer) {
   devServer.app.get('./app/**/*.html', function (req, res) {
     res.json({ custom: "response" });
   });
},
```

At that point there were no more compile errors for the server, but **_hot reload/hot module replacement_** wasn't working anymore. Whenever I changed the CSS it did a FULL PAGE RELOAD every time!

Then I spent another half hour combing the course Q&A for answers. I found one user, Simeon who said this:

> For those who are using Webpack 5 and webpack-dev-server >= 4.0.0.
> 
> In webpack.config.js change before option with the following:
> 
> watchFiles: ('./app/**/*.html'),

So, I removed onBeforeSetupMiddleware and replaced with `watchFiles`. Restarted the dev server. NO CHANGE --whenever CSS is changed it does a full page refresh now!

PS: If it helps, here's my GitHub repo where you can see my all my code:
https://github.com/codewizard13/travel-site/tree/ERROR_HotReload

**Comment by alexander-akait** [comment]
You should exclude files which support HMR from `watchFiles`, this option only for full browser reload, for example when developer change something in `php` files
